File: dane.pl
#!/usr/bin/env perl

use strict;
use warnings FATAL => 'all';


sub main {
    my ($host, $port, @openssl_args) = @_;

    my $cmdline = "echo | openssl s_client -servername ${host} -connect ${host}:${port} -showcerts @{openssl_args} 2>&1";
    my $cmd;
    open($cmd, "$cmdline|") || die "cannot exec $cmdline";
    my $pem;
    my $loadcert = 0;
    while (my $line = <$cmd>) {
        chomp $line;
        if ($line =~ /-----BEGIN CERTIFICATE-----/) {
            $loadcert = 1;
            $pem = "-----BEGIN CERTIFICATE-----\n";
        }
        elsif ($line =~ /-----END CERTIFICATE-----/) {
            $pem .= "-----END CERTIFICATE-----\n";
            $loadcert = 0;
            if (isCa($pem)) {
                analyze("--ca", "$pem", "$host", "$port");
            }
            else {
                analyze("", "$pem", "$host", "$port");
            }
        }
        elsif (defined $loadcert) {
            $pem .= "$line\n";
        }
    }
    close $cmd;
}

sub usage {
    die "usage: $0 <host> <port>\n";
}


sub isCa {
    my ($pem) = @_;

    my $cmdline = "echo \"$pem\" | openssl x509 -noout -text";
    my $cmd;
    open($cmd, "$cmdline|") || die "cannot exec $cmdline";
    while (my $line = <$cmd>) {
        chomp $line;
        if ($line =~ /CA:FALSE/) {
            return 0;
        }
        elsif ($line =~ /CA:TRUE/) {
            return 1;
        }
    }
    close $cmd;
}


sub analyze {
    my ($ca, $pem, $host, $port) = @_;
    my $cmdline = "echo \"$pem\" | openssl x509 -pubkey -noout";
    my $cmd;
    open($cmd, "$cmdline|") || die "cannot exec $cmdline";
    my $pubkeyFile = "/tmp/pubkey.pem.$$";
    open OUT, ">$pubkeyFile" || die "cannot open $pubkeyFile for write\n";
    while (my $line = <$cmd>) {
        chomp $line;
        print OUT "$line\n";
    }
    close OUT;
    close $cmd;

    $cmdline = "danetool --tlsa-rr $ca --host=${host} --port ${port} --proto=tcp --load-pubkey=${pubkeyFile}";
    open($cmd, "$cmdline|") || die "cannot exec $cmdline";
    while (my $line = <$cmd>) {
        chomp $line;
        print "$line\n";
    }
    close $cmd;
    unlink "$pubkeyFile";
}


my $host = shift;
my $port = shift;
my @openssl_args = @ARGV;

if (!defined($host) || !defined($port)) {
    usage();
}

main($host, $port, @openssl_args);

File: dane.py
#!/usr/bin/env python3

from subprocess import run, Popen, PIPE, STDOUT, TimeoutExpired
import sys, os
import re
import tempfile

def main(host, port, proto, openssl_args):
  cmd = [
    "openssl",
    "s_client",
    "-servername",
    host,
    "-connect",
    host + ":" + str(port),
    "-showcerts"
  ]
  for arg in openssl_args:
    cmd.append(arg)

  process = Popen(cmd, shell=False, stdin=PIPE, stdout=PIPE, stderr=PIPE)
  try:
    reply, err = process.communicate(b"\n", timeout=15)
    process.terminate()
    process_data(reply.decode("utf-8"), host, port, proto)
  except TimeoutExpired:
    proc.kill()
    reply, errs = proc.communicate()


def process_data(cert_data, host, port, proto):
  begin = re.compile('-----BEGIN CERTIFICATE-----')
  end = re.compile('-----END CERTIFICATE-----')
  pem = None
  loadcert = None
  for line in cert_data.splitlines():
    line = line.strip()
    if begin.match(line):
      loadcert = True
      pem = "-----BEGIN CERTIFICATE-----\n"
    elif end.match(line):
      loadcert = False
      pem += "-----END CERTIFICATE-----\n"
      analyze(isCa(pem), pem, host, port, proto)
    elif loadcert != None:
      pem += line + "\n";

def analyze(ca, pem, host, port, proto):
  cmd = [
    "openssl",
    "x509",
    "-noout",
    "--pubkey"
  ]
  pubkeyFile = None
  process = Popen(cmd, shell=False, stdin=PIPE, stdout=PIPE, stderr=PIPE)
  try:
    reply, err = process.communicate(str.encode(pem), timeout=15)
    process.terminate()
    fp = tempfile.NamedTemporaryFile(delete=False)
    pubkeyFile = fp.name
    fp.write(reply)
    fp.close()
  except TimeoutExpired:
    proc.kill()
    reply, errs = proc.communicate()

  cmd = [
    "danetool",
    "--tlsa-rr",
    "--host=" + host,
    "--port",
    str(port),
    "--proto=" + proto,
    "--load-pubkey=" + pubkeyFile
  ]
  if ca:
    cmd.append("--ca")
  process = Popen(cmd, shell=False, stdout=PIPE, stderr=PIPE)
  try:
    reply, err = process.communicate(timeout=15)
    process.terminate()
    if reply != None:
      print("RR: " + reply.decode("utf-8").strip())
  except TimeoutExpired:
    proc.kill()
    reply, errs = proc.communicate()
  os.remove(pubkeyFile)

def isCa(pem):
  cmd = [
    "openssl",
    "x509",
    "-noout",
    "-text"
  ]
  process = Popen(cmd, shell=False, stdin=PIPE, stdout=PIPE, stderr=PIPE)
  try:
    reply, err = process.communicate(str.encode(pem), timeout=15)
    process.terminate()
    ca_true = re.compile('CA:TRUE')
    ca_false = re.compile('CA:FALSE')
    for line in reply.decode("utf-8").splitlines():
      if ca_true.match(line.strip()):
        return True
      elif ca_false.match(line.strip()):
        return False
  except TimeoutExpired:
    proc.kill()
    reply, errs = proc.communicate()
    return False

host=sys.argv[1]
port=sys.argv[2]
args=sys.argv[3:]
main(host, port, "tcp", args)

